src/frontend/src/expr/mod.rs

```diff
@@ -176,18 +183,60 @@ impl ExprImpl {
         matches!(self, ExprImpl::Literal(literal) if literal.return_type() == DataType::Varchar)
     }
 
+    pub fn cast_implicit_or_shrink(self, target: DataType) -> Result<CastResult> {
+        let implicit_error = match self.clone().cast_implicit(target.clone()) {
+            Ok(expr) => return Ok(CastResult::Success(expr)),
+            Err(err) => err,
+        };
+        match self.cast_shrink(target) {
+            Ok(CastResult::Failed) | Err(_) => Err(implicit_error.into()),
+            ok => ok,
+        }
+    }
+
+    pub fn cast_shrink(self, target: DataType) -> Result<CastResult> {
+        match (self.return_type(), &target) {
+            (DataType::Int64, DataType::Int32)
+            | (DataType::Int64, DataType::Int16)
+            | (DataType::Int32, DataType::Int16) => self.cast_shrink_for_int(target),
+            _ => Ok(CastResult::Failed),
+        }
+    }
+
+    fn cast_shrink_for_int(self, target: DataType) -> Result<CastResult> {
+        let (upper_bound, lowwer_bound) = match (self.return_type(), &target) {
+            (DataType::Int64, DataType::Int32) => (i32::MAX as i64, i32::MIN as i64),
+            (DataType::Int64, DataType::Int16) | (DataType::Int32, DataType::Int16) => {
+                (i16::MAX as i64, i16::MIN as i64)
+            }
+            _ => return Ok(CastResult::Failed),
+        };
+        if let Some(scalar) = self.eval_row_const()? {
+            let value = scalar.as_integral();
```

Comment:
Hmm, will we be able to handle `float`, `decimal` as well?

Reply:
Yes

Reply:
Good question... So it is not enough to only check for minimum and maximum. For `where id < 2.5` and `where id > 2.5`, we need to round in different directions: `ceil` to `id < 3` and `floor` to `id > 2`. And when it is `id = 2.5`, this should evaluate to `false`.

There may be more complicated edge cases we are missing. Without optimization, the system always casts to a "larger" type and compare in that type. But in this optimization, we always want to compare in the type of order column, which may not be the "larger" type. Mimicking this `CMP_B(A_TO_B(x), A_TO_B(y))` via `CMP_A(B_TO_A(x), B_TO_A(y))` requires us to find the proper `B_TO_A` function given existing `A_TO_B`, `CMP_B` and `CMP_A` for all possible inputs.

More edge cases (PostgreSQL):
```
test=# select bigint '9223372036854775807' = double precision '9223372036854775806.5';
 ?column? 
----------
 t
(1 row)
test=# select (double precision '9223372036854775806.5')::bigint;
ERROR:  bigint out of range
test=# select (double precision '9223372036854775295')::bigint;
        int8         
---------------------
 9223372036854774784
(1 row)
```

Given it is an optimization, my suggestions is just to do it with best effort, and remain unoptimized for cases hard to handle correctly. int16/int32/int64 may be a good subset.

Also given its best-effort nature, the "cast_shrink" series helper functions should be limited to a mod for this special optimization, rather than on `ExprImpl`.